File: tests/test_table_model.py
"""表格模型测试 - 重点覆盖增量刷新与 hover 动画"""
import pytest
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from PyQt6.QtCore import Qt, QModelIndex

from gui.table_model import HistoryTableModel, NewRole, HoverRole


def _video_row(bvid: str, title: str = "标题", author: str = "UP主",
               view_time: str = "2024-06-20 12:00:00", category: str = "") -> dict:
    return {
        "类型": "video",
        "标题": title,
        "UP主": author,
        "BV号": bvid,
        "观看时间": view_time,
        "分类": category,
        "总时长(秒)": 600,
        "已观看(秒)": 300,
    }


class TestHistoryTableModel:
    """HistoryTableModel 单元测试"""

    def test_set_rows(self, qtbot):
        model = HistoryTableModel()
        with qtbot.wait_signal(model.modelReset):
            model.set_rows([_video_row("BV1"), _video_row("BV2")])
        assert model.rowCount() == 2

    def test_update_rows_insert_incrementally(self, qtbot):
        model = HistoryTableModel([_video_row("BV1")])
        assert model.rowCount() == 1

        with qtbot.wait_signal(model.rowsInserted):
            model.update_rows([_video_row("BV1"), _video_row("BV2")])
        assert model.rowCount() == 2
        assert model._rows[1]["BV号"] == "BV2"

    def test_update_rows_remove_incrementally(self, qtbot):
        model = HistoryTableModel([_video_row("BV1"), _video_row("BV2")])
        with qtbot.wait_signal(model.rowsRemoved):
            model.update_rows([_video_row("BV1")])
        assert model.rowCount() == 1
        assert model._rows[0]["BV号"] == "BV1"

    def test_update_rows_update_in_place(self, qtbot):
        model = HistoryTableModel([_video_row("BV1", title="旧标题")])
        with qtbot.wait_signal(model.dataChanged):
            model.update_rows([_video_row("BV1", title="新标题")])
        assert model.rowCount() == 1
        assert model._rows[0]["标题"] == "新标题"

    def test_update_rows_preserves_existing_identity(self, qtbot):
        """增量刷新应保持已有行的身份与顺序，排序由视图层 sortByColumn 负责"""
        model = HistoryTableModel([_video_row("BV1"), _video_row("BV2")])
        original_id = id(model._rows[0])
        model.update_rows([_video_row("BV1"), _video_row("BV2", title="更新")])
        assert id(model._rows[0]) == original_id
        assert model._rows[1]["标题"] == "更新"

    def test_set_new_ids_emits_data_changed(self, qtbot):
        model = HistoryTableModel([_video_row("BV1"), _video_row("BV2")])
        with qtbot.wait_signal(model.dataChanged):
            model.set_new_ids({"video_BV1"})
        assert model.data(model.index(0, 0), NewRole) is True
        assert model.data(model.index(1, 0), NewRole) is False

    def test_set_new_ids_no_reset(self, qtbot):
        model = HistoryTableModel([_video_row("BV1")])
        # modelReset 不应被触发
        with qtbot.assert_not_emitted(model.modelReset):
            model.set_new_ids({"video_BV1"})

    def test_hover_row_animation(self, qtbot):
        model = HistoryTableModel([_video_row("BV1")])
        model.set_hover_row(0)
        qtbot.wait(50)
        assert model._hover_row == 0
        assert 0.0 <= model._hover_progress <= 1.0
        assert model.data(model.index(0, 0), HoverRole) == model._hover_progress

        model.set_hover_row(-1)
        qtbot.wait(50)
        assert model._hover_row == -1

File: tests/gui/test_table_model.cpp
#include <gtest/gtest.h>

#include <QCoreApplication>

#include "core/models.h"
#include "gui/history_table_model.h"

using namespace bili;

class HistoryTableModelTest : public ::testing::Test {
protected:
    void SetUp() override {
        model = new gui::HistoryTableModel();
    }
    void TearDown() override {
        delete model;
    }
    gui::HistoryTableModel* model = nullptr;
};

static RecordPtr makeVideo(const QString& bv, const QString& title,
                           const QString& author, int percent) {
    auto r = std::make_shared<VideoRecord>();
    r->type = RecordType::Video;
    r->bvId = bv;
    r->bvid = bv;
    r->title = title;
    r->authorName = author;
    r->progressPercent = percent;
    r->viewAt = QDateTime::currentDateTime();
    return r;
}

TEST_F(HistoryTableModelTest, setRowsEmitsReset) {
    int resetCount = 0;
    QObject::connect(model, &gui::HistoryTableModel::modelReset, [&resetCount]() {
        ++resetCount;
    });

    RecordList rows;
    rows.push_back(makeVideo(QStringLiteral("BV1"), QStringLiteral("A"),
                             QStringLiteral("Up"), 10));
    model->setRows(rows);

    EXPECT_EQ(model->rowCount(), 1);
    EXPECT_EQ(model->columnCount(), 8);
    EXPECT_EQ(resetCount, 1);
}

TEST_F(HistoryTableModelTest, updateRowsInsertRemoveUpdate) {
    int inserted = 0;
    int removed = 0;
    int changed = 0;
    QObject::connect(model, &gui::HistoryTableModel::rowsInserted,
                     [&](const QModelIndex&, int, int) { ++inserted; });
    QObject::connect(model, &gui::HistoryTableModel::rowsRemoved,
                     [&](const QModelIndex&, int, int) { ++removed; });
    QObject::connect(model, &gui::HistoryTableModel::dataChanged,
                     [&](const QModelIndex&, const QModelIndex&) { ++changed; });

    RecordList rows;
    rows.push_back(makeVideo(QStringLiteral("BV1"), QStringLiteral("A"),
                             QStringLiteral("Up1"), 10));
    rows.push_back(makeVideo(QStringLiteral("BV2"), QStringLiteral("B"),
                             QStringLiteral("Up2"), 20));
    model->setRows(rows);

    RecordList rows2;
    rows2.push_back(makeVideo(QStringLiteral("BV2"), QStringLiteral("B"),
                              QStringLiteral("Up2"), 50));
    rows2.push_back(makeVideo(QStringLiteral("BV3"), QStringLiteral("C"),
                              QStringLiteral("Up3"), 30));
    model->updateRows(rows2);

    EXPECT_EQ(model->rowCount(), 2);
    EXPECT_EQ(inserted, 1);
    EXPECT_EQ(removed, 1);
    EXPECT_GE(changed, 1);

    EXPECT_EQ(model->data(model->index(0, 0)).toString(), QStringLiteral("视频"));
    EXPECT_EQ(model->data(model->index(0, 1)).toString(), QStringLiteral("B"));
    EXPECT_EQ(model->data(model->index(1, 1)).toString(), QStringLiteral("C"));
}

TEST_F(HistoryTableModelTest, setNewIdsAndHoverRow) {
    RecordList rows;
    rows.push_back(makeVideo(QStringLiteral("BV1"), QStringLiteral("A"),
                             QStringLiteral("Up"), 10));
    rows.push_back(makeVideo(QStringLiteral("BV2"), QStringLiteral("B"),
                             QStringLiteral("Up"), 20));
    model->setRows(rows);

    int changed = 0;
    QObject::connect(model, &gui::HistoryTableModel::dataChanged,
                     [&](const QModelIndex&, const QModelIndex&) { ++changed; });

    QSet<QString> newIds;
    newIds.insert(QStringLiteral("video:BV1"));
    model->setNewIds(newIds);
    EXPECT_GT(changed, 0);
    EXPECT_TRUE(model->data(model->index(0, 0), gui::HistoryTableModel::NewRole).toBool());
    EXPECT_FALSE(model->data(model->index(1, 0), gui::HistoryTableModel::NewRole).toBool());

    changed = 0;
    model->setHoverRow(1);
    EXPECT_GT(changed, 0);
    EXPECT_TRUE(model->data(model->index(1, 0), gui::HistoryTableModel::HoverRole).toBool());
    EXPECT_FALSE(model->data(model->index(0, 0), gui::HistoryTableModel::HoverRole).toBool());
}

TEST_F(HistoryTableModelTest, customRoles) {
    auto r = makeVideo(QStringLiteral("BV1"), QStringLiteral("A"),
                       QStringLiteral("Up"), 42);
    r->progress = QStringLiteral("12:34");
    RecordList rows;
    rows.push_back(r);
    model->setRows(rows);

    EXPECT_EQ(model->data(model->index(0, 0), gui::HistoryTableModel::TypeRole).toString(),
              QStringLiteral("视频"));
    EXPECT_EQ(model->data(model->index(0, 0), gui::HistoryTableModel::ProgressRole).toInt(), 42);
    EXPECT_EQ(model->data(model->index(0, 0), gui::HistoryTableModel::ProgressTextRole).toString(),
              QStringLiteral("12:34"));
    EXPECT_EQ(model->data(model->index(0, 0), gui::HistoryTableModel::LinkRole).toString(),
              QStringLiteral("video:BV1"));
}
